### PR TITLE
fix not updating the nav view when add/removing profiles

### LINKED ISSUE #15140: Adding or deleting a new profile to SUI no longer adds it to the nav list
### Windows Terminal version

1.18.96

### Windows build number

_No response_

### Other Software

_No response_

### Steps to reproduce

Add a new profile, by duplicating or making a net new one.

Delete an existing profile.

### Expected Behavior

_No response_

### Actual Behavior

It does not show up (or disappear.)

### PR DESCRIPTION
**Pull request by zadjii-msft**
* make the list of MenuItems observable, so the nav view can actually listen for changes to it
* Use the MenuItemsSource to find the index to add at, rather than the MenuItems (which isn't accurate anymore)
* Stash a single observable vector as the menuitemsource, and modify that whenever we need to do modifications.
  * I attempted to create a new vector, then copy into the new one, then replace the MenuItemsSource with the new vector, but that _refused_ to work. So let's just... not.

Regressed in #14630
Closes #15140

Manually validated that this and #13673 are still fixed

### CODE REVIEW

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 1]
```diff
@@ -666,9 +669,10 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         // remove selected item
         uint32_t index;
         auto selectedItem{ SettingsNav().SelectedItem() };
-        auto menuItems{ SettingsNav().MenuItems() };
+        auto menuItems{ SettingsNav().MenuItemsSource().try_as<Collections::IVector<IInspectable>>() };
         menuItems.IndexOf(selectedItem, index);
         menuItems.RemoveAt(index);
+        SettingsNav().MenuItemsSource(menuItems);
```
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:675]: erm null, not empty.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 2]
```diff
@@ -612,7 +612,10 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         const auto profileViewModel{ _viewModelForProfile(newProfile, _settingsClone) };
         profileViewModel.SetupAppearances(_colorSchemesPageVM.AllColorSchemes());
         const auto navItem{ _CreateProfileNavViewItem(profileViewModel) };
-        SettingsNav().MenuItems().InsertAt(index, navItem);
+
+        auto navItems{ SettingsNav().MenuItemsSource().try_as<Collections::IVector<IInspectable>>() };
+        navItems.InsertAt(index, navItem);
+        SettingsNav().MenuItemsSource(navItems);
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsEditor/MainPage.cpp:618]: what's the purpose of making it observable if we're also resetting it? this mixes the metaphors...

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:675]: it absolutely shouldn't come out null. The first thing we do is make this a vector - I could guard against this though it does seem like it'd hide any issues if we changed it in the future

**Review by DHowett — APPROVED** [review]
Does this also fix the issue where Save/Discard lose track of what profile you're looking at?

**Comment by DHowett** [comment]
uh, this is weird.

now discard/save bring you back to the Startup page.

and when I discard, my first profile disappears and is replaced with a hole:

image

**Review by DHowett — CHANGES_REQUESTED** [review]
above

**Comment by DHowett** [comment]
I love that I can pull the unpackaged distribution from your build and just run it as-is

**Review by DHowett — APPROVED** [review]
Tested scenarios:

- [x] new profile
- [x] discard with selection
- [x] save with selection
- [x] save json outside of app
- [x] delete profile
- [x] delete profile in json while it is selected

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 3]
```diff
@@ -117,54 +117,19 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
...
+        // By manipulating a MenuItemsSource this way, rather than manipulating the
+        // MenuItems directly, we avoid a crash in WinUI.
+        //
+        // By making the vector only _originalNumItems big to start, GetMany
+        // will only fill that number of elements out of the current source.
+        std::vector<IInspectable> menuItemsSTL(_originalNumItems, nullptr);
+        _menuItemSource.GetMany(0, menuItemsSTL);
+        _menuItemSource.Clear();
+        // now, just stick them back in.
+        _menuItemSource.ReplaceAll(menuItemsSTL);
```
[the remark this answers is not in the document]
Comment (lhecker) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:132]: Same question here.

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 4]
```diff
@@ -251,8 +217,10 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
     {
         uint32_t insertIndex;
         auto selectedItem{ SettingsNav().SelectedItem() };
-        auto menuItems{ SettingsNav().MenuItems() };
-        menuItems.IndexOf(selectedItem, insertIndex);
+        if (_menuItemSource)
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsEditor/MainPage.cpp:220]: It can be null?

File: src/cascadia/TerminalSettingsEditor/MainPage.cpp [hunk 5]
```diff
@@ -612,7 +581,11 @@ namespace winrt::Microsoft::Terminal::Settings::Editor::implementation
         const auto profileViewModel{ _viewModelForProfile(newProfile, _settingsClone) };
         profileViewModel.SetupAppearances(_colorSchemesPageVM.AllColorSchemes());
         const auto navItem{ _CreateProfileNavViewItem(profileViewModel) };
-        SettingsNav().MenuItems().InsertAt(index, navItem);
+
+        if (const auto& navItems{ _menuItemSource })
+        {
+            navItems.InsertAt(index, navItem);
+        }
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsEditor/MainPage.cpp:588]: Why not just this, like in the other places?
```suggestion
        if (_menuItemSource)
        {
            _menuItemSource.InsertAt(index, navItem);
        }
```

Same question here btw: Can it be null?

Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsEditor/MainPage.cpp:588]: a bad string replace, that's why

**Pull request merged**